go/dbagent/dbagent.go

```diff
@@ -0,0 +1,360 @@
+	query := `SELECT user FROM mysql.user WHERE user=?;`
+	err := QueryData(query, sqlutils.Args(username), func(m sqlutils.RowMap) error {
+		UserExists = (len(m.GetString("user")) != 0)
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return UserExists
+}
+
+func HasGrant(username string, grant string) (HasGrant bool) {
+	query := `SELECT * FROM mysql.user WHERE user=?;`
+	err := QueryData(query, sqlutils.Args(username), func(m sqlutils.RowMap) error {
+		HasGrant = (m.GetString(grant) == "Y")
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return HasGrant
+}
+
+func CreateUser(username string, host string, password string) error {
+	_, err := ExecuteQuery(`CREATE USER IF NOT EXISTS ?@? IDENTIFIED BY ?;`, username, host, password)
+	return err
+}
+
+func GrantUser(username string, host string, grant string, object string, password string) error {
+	var query string
+	if len(password) > 0 {
+		query = fmt.Sprintf("GRANT %s ON %s TO '%s'@'%s' IDENTIFIED BY '%s';", grant, object, username, host, password)
+	} else {
+		query = fmt.Sprintf("GRANT %s ON %s TO '%s'@'%s';", grant, object, username, host)
+	}
+	_, err := ExecuteQuery(query)
+	return err
+}
+
+func ManageReplicationUser() error {
+	var err error
+	version, err := GetMySQLVersion()
+	if err != nil {
+		log.Errore(err)
+	}
+	if !UserExists(config.Config.MySQLReplicationUser) {
+		if version == "5.7" {
+			err = CreateUser(config.Config.MySQLReplicationUser, "%", config.Config.MySQLReplicationPassword)
+			if err != nil {
+				return err
+			}
+		} else {
+			err = GrantUser(config.Config.MySQLReplicationUser, "%", "REPLICATION SLAVE", "*.*", config.Config.MySQLReplicationPassword)
+		}
+	}
+	if !HasGrant(config.Config.MySQLReplicationUser, "Repl_slave_priv") {
+		err = GrantUser(config.Config.MySQLReplicationUser, "%", "REPLICATION SLAVE", "*.*", "")
+	}
+	query := fmt.Sprintf(`FLUSH PRIVILEGES;`)
+	_, err = ExecuteQuery(query)
+	if err != nil {
+		return log.Errore(err)
+	}
+	return err
+}
+
+func GenerateBackupForUsers(users []string) (backup string, err error) {
+	if db, err := OpenConnection(); err == nil {
+		for _, user := range users {
+			// small hack cos MySQL 5.6 doesn't have DROP USER IF EXISTS. USAGE is synonym for "no privileges"
+			backup += fmt.Sprintf("GRANT USAGE ON *.* TO %s IDENTIFIED BY 'temporary_password'; DROP USER %s;", user, user)
+			query := fmt.Sprintf("SHOW CREATE USER %s", user)
+			res, err := sqlutils.QueryResultData(db, query)
+			if err != nil {
+				log.Errore(err)
+			} else {
+				backup += res[0][0].String + ";\n"
+			}
+			query = fmt.Sprintf("SHOW GRANTS FOR %s", user)
+			res, err = sqlutils.QueryResultData(db, query)
+			if err != nil {
+				log.Errore(err)
+			} else {
+				for r := range res {
+					backup += res[r][0].String + ";\n"
+				}
+			}
+		}
+	}
+	return backup, err
+}
+
+func StartSlave(sourceHost string, sourcePort int, logFile string, position string, gtidPurged string) error {
```

Comment:
Doesn't this re-implement functionality in `orchestrator`?

Reply:
Agree. Will create API `/api/backup-metadata/:seedId/:seedMethod/:backupFolder` which will return log position\file and gtid_purged if it exists (`{"BinlogCoordinates":{"LogFile":"mysql-bin.000008","LogPos":194,"Type":0},"GTIDPurged":""}`) and use 
`func ResetMaster(instanceKey *InstanceKey) (*Instance, error)`, 
`func setGTIDPurged(instance *Instance, gtidPurged string)`, 
`func ChangeMasterTo(instanceKey *InstanceKey, masterKey *InstanceKey, masterBinlogCoordinates *BinlogCoordinates, skipUnresolve bool, gtidHint OperationGTIDHint) (*Instance, error)`,
`func StartSlave(instanceKey *InstanceKey)`
on the orchestrator side to configure and start replication. 
Replace this with:
`func SetReplicationUserAndPassword() error {
	query := fmt.Sprintf(`CHANGE MASTER TO MASTER_USER='%s', MASTER_PASSWORD='%s';`,
		config.Config.MySQLReplicationUser, config.Config.MySQLReplicationPassword)
	_, err := ExecuteQuery(query)
	return err
}`
as orchestrator may not have information about replication user in babfa38